EIPS/eip-1605.md

```diff
@@ -0,0 +1,404 @@
+
+Given an event name and series of event parameters, we split them into two sub-series: those which are indexed and those which are not. Those which are indexed, which may number up to 3, are used alongside the Keccak hash of the event signature to form the topics of the log entry. Those which are not indexed form the byte array of the event.
+
+In effect, a log entry using this ABI is described as:
+
+- ``address``: the address of the contract (intrinsically provided by Ethereum);
+- ``topics[0]``: ``keccak(EVENT_NAME+"("+EVENT_ARGS.map(canonical_type_of).join(",")+")")`` (``canonical_type_of`` is a function that simply returns the canonical type of a given argument, e.g. for ``uint indexed foo``, it would return ``uint256``). If the event is declared as ``anonymous`` the ``topics[0]`` is not generated;
+- ``topics[n]``: ``EVENT_INDEXED_ARGS[n - 1]`` (``EVENT_INDEXED_ARGS`` is the series of ``EVENT_ARGS`` that are indexed);
+- ``data``: ``abi_serialise(EVENT_NON_INDEXED_ARGS)`` (``EVENT_NON_INDEXED_ARGS`` is the series of ``EVENT_ARGS`` that are not indexed, ``abi_serialise`` is the ABI serialisation function used for returning a series of typed values from a function, as described above).
+
+For all fixed-length Solidity types, the ``EVENT_INDEXED_ARGS`` array contains the 32-byte encoded value directly. However, for *types of dynamic length*, which include ``string``, ``bytes``, and arrays, ``EVENT_INDEXED_ARGS`` will contain the *Keccak hash* of the packed encoded value, rather than the encoded value directly. This allows applications to efficiently query for values of dynamic-length types (by setting the hash of the encoded value as the topic), but leaves applications unable to decode indexed values they have not queried for. For dynamic-length types, application developers face a trade-off between fast search for predetermined values (if the argument is indexed) and legibility of arbitrary values (which requires that the arguments not be indexed). Developers may overcome this tradeoff and achieve both efficient search and arbitrary legibility by defining events with two arguments — one indexed, one not — intended to hold the same value.
+
+## Acknowledgements
+
+This specification has its origin in the [Ethereum Wiki: Ethereum Contract ABI](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI). Since June 2017 it was maintained in the [Solidity repository](https://github.com/ethereum/solidity/blob/develop/docs/abi-spec.rst).
+
+There are numerous contributors to the specification. This EIP only lists the major contributors on the top. Please review the change history of the above two links for clarity.
+
+## Copyright
+
+This was imported from Solidity ([973d91ec](https://github.com/ethereum/solidity/blob/973d91ec894a5438597855ffb9e53f0a20c18696/docs/abi-spec.rst)). Solidity is licensed under [GNU GPL-3](https://www.gnu.org/licenses/gpl-3.0.en.html) and as an extension this document as well.
```

Comment:
Policy is that all EIPs be CC0.  If there is *any* way to license this as CC0 that is preferable to re-opening the massive holy war that is EIP licensing again.